src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs

```diff
@@ -101,11 +148,18 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS OnAsyncCompletion(IntPt
             return NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_PENDING;
         }
 
-        private static void CompleteRequest(IISHttpContext context, Task<bool> completedTask)
+        private static void CompleteRequest(IISHttpContext context, IISHttpServer server, Task<bool> completedTask)
         {
             // Post completion after completing the request to resume the state machine
             context.PostCompletion(ConvertRequestCompletionResults(completedTask.Result));
 
+            if (Interlocked.Decrement(ref server._outstandingRequests) == 0 && server.Stopping)
+            {
+                // All requests have been drained.
+                NativeMethods.http_stop_calls_into_mananged();
```

Comment:
You also need to call this on line 92 if there are no active requests when you initiate shutdown.